src/v/kafka/server/replicated_partition.cc

```diff
@@ -475,11 +475,19 @@ ss::future<error_code> replicated_partition::validate_fetch_offset(
         if (fetch_offset < start_offset()) {
             ec = error_code::offset_out_of_range;
         } else if (
-          fetch_offset > high_watermark()
-          && fetch_offset <= leader_high_watermark()) {
-            ec = error_code::offset_not_available;
-        } else if (fetch_offset > leader_high_watermark()) {
+          fetch_offset > log_end_offset()
+          && fetch_offset > leader_high_watermark()) {
```

Comment:
nit: this is a bit hard to read because conceptually these conditionals go left to right then to the left again in the offset space. IMO it is better to rewrite it this way:
```
if (fetch_offset > leader_high_watermark()) {
    if (fetch_offset <= log_end_offset()) {
        ec = error_code::offset_not_available;
    } else {
        ec = error_code::offset_out_of_range;
    }
}
```